src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -343,6 +343,139 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
         appendUnusedGroupByColumn(select_query, source_columns);
 }
 
+///recursive traversal and check for optimizeAggregateFunctionsOfGroupByKeys
+struct KeepAggregateFunctionMatcher
```

Comment:
Move this and other Matchers/Visitors out of SyntaxAnalyzer.h into own file: `<...>Visitor.h`.